scope.go

```diff
@@ -0,0 +1,622 @@
+package rcmgr
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/libp2p/go-libp2p-core/network"
+)
+
+// resources tracks the current state of resource consumption
+type resources struct {
+	limit Limit
+
+	nconnsIn, nconnsOut     int
+	nstreamsIn, nstreamsOut int
+	nfd                     int
+
+	memory int64
+}
+
+// A resourceScope can be a DAG, where a downstream node is not allowed to outlive an upstream node
+// (ie cannot call Done in the upstream node before the downstream node) and account for resources
+// using a linearized parent set.
+// A resourceScope can be a txn scope, where it has a specific owner; txn scopes create a tree rooted
+// at the owner (which can be a DAG scope) and can outlive their parents -- this is important because
+// txn scopes are the main *user* interface for memory management, and the user may call
+// Done in a txn scope after the system has closed the root of the txn tree in some background
+// goroutine.
+// If we didn't make this distinction we would have a double release problem in that case.
+type resourceScope struct {
+	sync.Mutex
+	done   bool
+	refCnt int
+
+	rc          resources
+	owner       *resourceScope   // set in transaction scopes, which define trees
+	constraints []*resourceScope // set in DAG scopes, it's the linearized parent set
+
+	name string // for debugging purposes
+}
+
+var _ network.ResourceScope = (*resourceScope)(nil)
+var _ network.TransactionalScope = (*resourceScope)(nil)
+
+func newResourceScope(limit Limit, constraints []*resourceScope, name string) *resourceScope {
+	for _, cst := range constraints {
+		cst.IncRef()
+	}
+	return &resourceScope{
+		rc:          resources{limit: limit},
+		constraints: constraints,
+		name:        name,
+	}
+}
+
+func newTxnResourceScope(owner *resourceScope) *resourceScope {
+	return &resourceScope{
+		rc:    resources{limit: owner.rc.limit},
+		owner: owner,
+		name:  fmt.Sprintf("%s.txn", owner.name),
+	}
+}
+
+// Resources implementation
+func (rc *resources) checkMemory(rsvp int64, prio uint8) error {
+	// overflow check; this also has the side effect that we cannot reserve negative memory.
+	newmem := rc.memory + rsvp
+	limit := rc.limit.GetMemoryLimit()
+	threshold := (1 + int64(prio)) * limit / 256
+
+	if newmem > threshold {
+		return network.ErrResourceLimitExceeded
+	}
+
+	return nil
+}
+
+func (rc *resources) reserveMemory(size int64, prio uint8) error {
+	if err := rc.checkMemory(size, prio); err != nil {
+		return err
+	}
+
+	rc.memory += size
+	return nil
+}
+
+func (rc *resources) releaseMemory(size int64) {
+	rc.memory -= size
+
+	// sanity check for bugs upstream
+	if rc.memory < 0 {
+		panic("BUG: too much memory released")
```

Comment:
Age-old question here: are we sure we want to panic? In most cases memory will be dynamically calculated; panicking is fine for programming errors that can be caught by static code analysis (i.e. there is tooling to prevent the developer from making a mistake), but this isn't one of those. Panicking here is quite unforgiving here, specially when then recovery path is obvious: reset the value back to zero and log a warning.

Reply:
ok, we can do that.

Reply:
it logs an error now and resets.